odoo/tools/misc.py

```diff
@@ -1821,6 +1821,64 @@ def verify_hash_signed(env, scope, payload):
     return None
 
 
+def limited_field_access_token(record, field_name, timestamp=None):
+    """Generate a token granting access to the given record and field_name from
+    the binary routes (/web/content or /web/image).
+
+    The validitiy of the token is determined by the timestamp parameter.
+    When it is not specified, a timestamp is automatically generated with a
+    validity of at least 14 days. For a given record and field_name, the
+    generated timestamp is deterministic within a 14-day period (even across
+    different days/months/years) to allow browser caching, and expires after
+    maximum 42 days to prevent infinite access. Different record/field
+    combinations expire at different times to prevent thundering herd problems.
+
+    :param record: the record to generate the token for
+    :type record: class:`odoo.models.Model`
+    :param field_name: the field name of record to generate the token for
+    :type field_name: str
+    :param timestamp: expiration timestamp of the token, or None to generate one
+    :type timestamp: int, optional
+    :return: the token, which includes the timestamp in hex format
+    :rtype: string
+    """
+    record.ensure_one()
+    if not timestamp:
+        now = datetime.datetime.now()
+        epoch = datetime.datetime(1970, 1, 1)
+        start_of_period = epoch + datetime.timedelta(days=((now - epoch).days // 14) * 14)
+        unique_str = repr((record._name, record.id, field_name))
+        days_offset = int.from_bytes(hashlib.sha1(f"d-{unique_str}".encode()).digest(), "little")
+        seconds_offset = int.from_bytes(hashlib.sha1(f"s-{unique_str}".encode()).digest(), "little")
+        date = start_of_period + datetime.timedelta(
+            days=28 + days_offset % 14, seconds=seconds_offset % (24 * 60 * 60)
+        )
+        timestamp = hex(int(date.timestamp()))
```

Comment:
let's remove the `0x` hex adds, so all chars are [0-9a-f]
```suggestion
        timestamp = hex(int(date.timestamp())).removeprefix('0x')
```

btw `int("0xf", 16)` and `int("f", 16)` both work

Reply:
Good catch, the goal of having hex in the first place was to keep `[0-9a-f]` format for the whole token, so let's do it fully, otherwise just `-{timestamp}` would have been easier.

Reply:
Well almost all of the chars, we still have the `o` separator but I'm not sure if we can safely remove it.